Add CD guide for releasing plugins (#4341)
jenkins-infra/jenkins.io: content/doc/developer/publishing/releasing-cd.adoc

. Enabled the continuous delivery flag in link:https://github.com/jenkins-infra/repository-permissions-updater/[repository permission updater] (RPU) for your plugin.

[source,yaml]
----
cd:
  enabled: true
----

Once that has been merged, start checking https://github.com/jenkinsci/your-plugin/settings/secrets/actions and you should soon see `MAVEN_TOKEN` and `MAVEN_USERNAME` appear under Repository secrets.

== Release notes

Next, if you already have Release Drafter configured, remove any tag-template override in `.github/release-drafter.yml`, and delete `.github/workflows/release-drafter.yml` if using GitHub Actions (or remove the app from the repo otherwise). If you have not yet configured Release Drafter, just create `.github/release-drafter.yml` containing only:

[source,yaml]
----
_extends: .github
----

== Adding the GitHub action

Create `.github/workflows/cd.yaml` as a copy of the link:https://github.com/jenkinsci/.github/blob/master/workflow-templates/cd.yaml[cd.yaml template].

If this is your first time using a GitHub action on this project then navigate to the 'Actions' tab 

> in your GitHub repository and enable GitHub actions.

Author: pretty sure recently I had to do this. Happy to be proved wrong.

If you have a `.github/dependabot.yml`, it is a good idea to add:

[source,yaml]
----
- package-ecosystem: github-actions
  directory: /
  schedule:
    interval: daily
----

== Pom modifications

> For a regular component whose version number is not that meaningful:

Reviewer: Will need to call attention to the fact that you should have no expectation of explicitly picking version numbers in this case; was a source of confusion in the case of `token-macro`.

Author: is this what you meant? https://github.com/jenkins-infra/jenkins.io/pull/4341/files#r629353070

Reviewer: Works.

> [source,diff]

Reviewer: Maybe we should create a new goal like `incrementalify` (or option to that goal) to do this sort of edit. And/or a flag in `archetypes`?

Author: yes that would be great, I was thinking about this at least in context of the IRC bot, is there any reason not to automatically opt new plugins in to the cd.enabled flag? i.e. don't bother adding a flag to Jira in the hosting issue.

Also do we need a new flag in archetypes or can we assume plugins use the is hosted on jenkinsci flag should use this setup?

Reviewer: > is there any reason not to automatically opt new plugins in to the cd.enabled flag?

That would be a question for @daniel-beck probably.

> can we assume plugins us[ing] the is hosted on jenkinsci flag should use this setup?

We should probably get some more diverse adoption before making this the default for new plugins.

Reviewer 2: > is there any reason not to automatically opt new plugins in to the cd.enabled flag?

Assuming this is asking "why not set this even if unused":

* This effectively changes the permission model; anyone with commit access to the repo can deploy (even if jumping through hoops to get the token). Right now folks can add whoever they want to GH as committers and it doesn't affect the bits we distribute; `cd.enabled` changes this.
* Every repo adds about 1-2 seconds to the execution time of the RPU run.

----
--- a/.mvn/maven.config
+++ b/.mvn/maven.config
@@ -1,2 +1,3 @@
 -Pconsume-incrementals
 -Pmight-produce-incrementals
+-Dchangelist.format=%d.v%s
--- a/pom.xml
+++ b/pom.xml
@@ -7,7 +7,7 @@
-    <version>${revision}${changelist}</version>
+    <version>${changelist}</version>
     <packaging>hpi</packaging>
@@ -26,8 +26,7 @@
     <properties>
-        <revision>1.23</revision>
-        <changelist>-SNAPSHOT</changelist>
+        <changelist>999999-SNAPSHOT</changelist>
         <jenkins.version>2.176.4</jenkins.version>
     </properties>
----

> Here a CI/release build (-Dset.changelist specified) will be of the form `123.vabcdef456789`. A snapshot build will be `999999-SNAPSHOT`: arbitrary but treated as a snapshot by Maven and newer than any release.

Author: ```suggestion
Here a CI/release build (`-Dset.changelist` specified) will be of the form `123.vabcdef456789`. A snapshot build will be `999999-SNAPSHOT`: arbitrary but treated as a snapshot by Maven and newer than any release.
```

(e.g. enhancement rather than chore), your component will be released to Artifactory and 
release notes published in GitHub.
You do not need any special credentials or local checkout; just merge pull requests with suitable titles and labels.

You can also trigger a deployment explicitly, if the current commit has a passing check from Jenkins. Visit https://github.com/jenkinsci/your-plugin/actions?query=workflow%3Acd and click Run workflow.
If you prefer to only deploy explicitly, not on every push, just comment out the check_run section in the workflow.

== Fallback

You can also release manually if you have configurated your machine for link:../releasing-manually[manual release].
To cut a release:

[source,shell]
----
git checkout master
git pull --ff-only
mvn -Dset.changelist \
  -DaltDeploymentRepository=maven.jenkins-ci.org::default::https://repo.jenkins-ci.org/releases/ \
  clean deploy
----

== Troubleshooting

Check that `MAVEN_TOKEN` and `MAVEN_USERNAME` appear under Repository secrets.

=== The upload to the Maven repository fails with "401 Unauthorized"

Unauthorized means that the credentials were invalid, or not sent by Maven.

> This normally means that the secrets configured in the repository have expired, contact the infrastructure team in #jenkins-infra on link:https://freenode.net[Freenode].

Author: ```suggestion
This normally means that the secrets configured in the repository have expired, contact the infrastructure team in #jenkins-infra on link:https://freenode.net[Freenode].

Alternatively you can temporarily update the secrets yourself with your own personal credentials.
```

=== Further troubleshooting help

> If none of the provided solutions help, send an email to the link:/mailing-lists[Jenkins developers mailing list] and explain what you did, and how it failed.

Reviewer 3: While I agree that it will be faster to ask for help on IRC, creating a ticket will be useful for documentation purposes.

> If none of the provided solutions help, send an email to the link:/mailing-lists[Jenkins developers mailing list] and explain what you did, and how it failed.

Author: better?

_extends: .github
----

== Adding the GitHub action

Create `.github/workflows/cd.yaml` as a copy of the link:https://github.com/jenkinsci/.github/blob/master/workflow-templates/cd.yaml[cd.yaml template].

If this is your first time using a GitHub action on this project then navigate to the 'Actions' tab 
in your GitHub repository and enable GitHub actions.

If you have a `.github/dependabot.yml`, it is a good idea to add:

[source,yaml]
----
- package-ecosystem: github-actions
  directory: /
  schedule:
    interval: daily
----

== Pom modifications

For a regular component whose version number is not that meaningful:

[source,diff]
----
--- a/.mvn/maven.config
+++ b/.mvn/maven.config
@@ -1,2 +1,3 @@
 -Pconsume-incrementals
 -Pmight-produce-incrementals
+-Dchangelist.format=%d.v%s
--- a/pom.xml
+++ b/pom.xml
@@ -7,7 +7,7 @@
-    <version>${revision}${changelist}</version>
+    <version>${changelist}</version>
     <packaging>hpi</packaging>
@@ -26,8 +26,7 @@
     <properties>
-        <revision>1.23</revision>
-        <changelist>-SNAPSHOT</changelist>
+        <changelist>999999-SNAPSHOT</changelist>
         <jenkins.version>2.176.4</jenkins.version>
     </properties>
----

> Here a CI/release build (`-Dset.changelist` specified) will be of the form `123.vabcdef456789`. A snapshot build will be `999999-SNAPSHOT`: arbitrary but treated as a snapshot by Maven and newer than any release.

Author: ```suggestion
Here a CI/release build (`-Dset.changelist` specified) will be of the form `123.vabcdef456789`. A snapshot build will be `999999-SNAPSHOT`: arbitrary but treated as a snapshot by Maven and newer than any release.

It's worth communicating this to your users, as they will see a very different version number format than before.
The best way to do this is to add a line to the release notes, link:https://github.com/jenkinsci/azure-artifact-manager-plugin/releases/tag/86.va2aa4b1038c7[example note].
```

---
title: Setting up automated plugin release
layout: developerguide
references:
- url: /jep/229
  title: 'JEP-229: Continuous Delivery of Jenkins Components and Plugins'
---

> 

Reviewer 4: ```suggestion

WARNING: Continuous Delivery of Jenkins Components and Plugin (jep:229[]) is currently in preview.
See the referenced Jenkins enhancement proposal and references in it to see the current state.
Any feedback from early adopters will be appreciated.
```

Author: where is it marked as preview?

Reviewer: ```suggestion

NOTE: Continuous Delivery of Jenkins Components and Plugin (jep:229[]) is relatively new.
Any feedback from early adopters will be appreciated.
```

Reviewer 4: Roadmap-wise, an initiative is in preview until the respective JEP is formally accepted

Author: https://github.com/jenkinsci/jep/tree/master/jep/14 does not state that as far as I can tell.

I don't think this discussion is for this pull request.

Nothing I can see anywhere states that this requires a preview disclaimer, I think the current disclaimer is appropriate.

Please take the bits about github actions long term to the mailing list as it doesn't belong in this PR

Reviewer 4: FTR the public roadmap had a preview disclaimer until formal publishing and JEP acceptance approval by the Jenkins Governance Meeting and then by the BDFL delegate. See https://github.com/jenkins-infra/jenkins.io/pull/3527 for the pull request which removed the disclaimer

Author: That may be true but does not appear to be required?

(e.g. enhancement rather than chore), your component will be released to Artifactory and 
release notes published in GitHub.
You do not need any special credentials or local checkout; just merge pull requests with suitable titles and labels.

You can also trigger a deployment explicitly, if the current commit has a passing check from Jenkins. Visit https://github.com/jenkinsci/your-plugin/actions?query=workflow%3Acd and click Run workflow.
If you prefer to only deploy explicitly, not on every push, just comment out the check_run section in the workflow.

== Fallback

You can also release manually if you have configured your machine for link:../releasing-manually[manual release].
To cut a release:

[source,shell]
----
git checkout master
git pull --ff-only
mvn -Dset.changelist \
  -DaltDeploymentRepository=maven.jenkins-ci.org::default::https://repo.jenkins-ci.org/releases/ \
  clean deploy
----

== Troubleshooting

Check that `MAVEN_TOKEN` and `MAVEN_USERNAME` appear under Repository secrets.

=== The upload to the Maven repository fails with "401 Unauthorized"

Unauthorized means that the credentials were invalid, or not sent by Maven.

> This normally means that the secrets configured in the repository have expired, contact the infrastructure team in #jenkins-infra on link:https://freenode.net[Freenode].

Author: ```suggestion
This normally means that the secrets configured in the repository have expired, create an issue in the INFRA project on link:https://issues.jenkins.io/[Jira], and let the team know in #jenkins-infra on link:https://freenode.net[Freenode].
```

Here a CI/release build (`-Dset.changelist` specified) will be of the form `123.vabcdef456789`. A snapshot build will be `999999-SNAPSHOT`: arbitrary but treated as a snapshot by Maven and newer than any release.

It's worth communicating this to your users, as they will see a very different version number format than before.

> The best way to do this is to add a line to the release notes, link:https://github.com/jenkinsci/azure-artifact-manager-plugin/releases/tag/86.va2aa4b1038c7[example note].

Reviewer: ```suggestion
The best way to do this is to add a line to the release notes: link:https://github.com/jenkinsci/azure-artifact-manager-plugin/releases/tag/86.va2aa4b1038c7[example note].
```
